### 1. Import packages

In [1]:
import os
import sys

module_path = os.path.abspath(os.path.join(".."))
if module_path not in sys.path:
    sys.path.append(module_path)

Importing required modules and setting random seed

In [2]:
from k_nearest_neighbors_classifier import KNearestNeighborsClassifier
from sklearn import datasets
from sklearn import model_selection as ms
from sklearn.metrics import f1_score
from sklearn.neighbors import KNeighborsClassifier

from utils.scaler import StandardScaler

SEED = 42

### 2. Data preparation

#### 2.1 Loading Iris dataset

I'm going to use only first two features for now

In [3]:
X, y = datasets.load_iris(return_X_y=True)
X = X[:, :2]

#### 2.2 Splitting the data

In [4]:
X_train, X_test, y_train, y_test = ms.train_test_split(X, y, train_size=0.8, random_state=SEED)

#### 2.3 Scaling the data

In [5]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

### 3. Predictive modelling

#### 3.1 My implementation of kNN

In [6]:
knn = KNearestNeighborsClassifier(n_neighbors=5, metric="minkowski", p=4)

knn.fit(X_train_scaled, y_train)
pred = knn.predict(X_test_scaled)

f1 = f1_score(y_true=y_test, y_pred=pred, average="weighted").round(2)
print(f"F1 is equal to {f1}")

F1 is equal to 0.83


#### 3.2 Sklearn's implementation of kNN

In [7]:
knn_sklearn = KNeighborsClassifier(n_neighbors=5, metric="minkowski", p=4)
knn_sklearn.fit(X_train_scaled, y_train)
pred_sklearn = knn_sklearn.predict(X_test_scaled)

f1_sklearn = f1_score(y_true=y_test, y_pred=pred_sklearn, average="weighted").round(2)
print(f"F1 is equal to {f1_sklearn}")

F1 is equal to 0.8


### 4. Idea of kNN

The idea of kNN classifier is quite simple. Basic brute force implementation doesn't require any
training and works when the inference happens.
Let's describe it in a few steps:
1. Standardization of dataset. 
Scaled data should have zero mean and unit variance. (Write about importance of standardization)
2. For every example in test dataset: <br>
2.1 Calculate distance between example and training dataset (Write about different metrics)<br>
2.2 Sort the distances in increasing order <br>
2.3 Find the first <code>k</code> closest items (from training dataset) and remember their labels <br>
2.4 Use the labels to do the majority voting and find the most common class among the labels <br>
2.5 Return the majority class as a prediction for input example <br>

As we can see from the results, my brute force implementation of kNN classifier (with the same parameters) shows even better performance on a small, simple dataset as implementation from sklearn: F1 score = 0.83 (if I'm not mistaken somewhere). 

### TODO:
1. Investigate Nearest Neighbor Algorithms (`ball_tree`, `kd_tree`)
2. Investigate Nearest Neighbors Regression
3. Investigate effect of data standardization of the knn (possibly on k-means and linear models too)
4. Plot dependency between `p` for `minkowski` distance, `n_neighbors` and F1 score for iris data
5. Investigate ANN approaches (Annoy, N2, hnswlib, etc.)